src/locator.rs

```diff
@@ -0,0 +1,643 @@
+            };
+            assert_eq!(
+                parsed.revision().as_ref().map(|r| r.as_str()),
+                revision,
+                "'revision' in '{input}' must match",
+            );
+        }
+    }
+
+    #[test]
+    fn render_with_org() {
+        let locator = Locator::builder()
+            .fetcher(Fetcher::Custom)
+            .org_id(1234)
+            .project("foo/bar")
+            .revision("123abc")
+            .build();
+
+        let rendered = locator.to_string();
+        assert_eq!("custom+1234/foo/bar$123abc", rendered);
+
+        let package_only = locator.into_package();
+        let rendered = package_only.to_string();
+        assert_eq!("custom+1234/foo/bar", rendered);
+    }
+
+    #[test]
+    fn render_with_revision() {
+        let locator = Locator::builder()
+            .fetcher(Fetcher::Custom)
+            .project("foo/bar")
+            .revision("123abc")
+            .build();
+
+        let rendered = locator.to_string();
+        assert_eq!("custom+foo/bar$123abc", rendered);
+    }
+
+    #[test]
+    fn render_project() {
+        let locator = Locator::builder()
+            .fetcher(Fetcher::Custom)
+            .project("foo/bar")
+            .build();
+
+        let rendered = locator.to_string();
+        assert_eq!("custom+foo/bar", rendered);
+    }
+
+    #[test]
+    fn roundtrip_serialization() {
+        let input = Locator::builder()
+            .fetcher(Fetcher::Custom)
+            .project("foo")
+            .revision("bar")
+            .org_id(1)
+            .build();
+
+        let serialized = serde_json::to_string(&input).expect("must serialize");
+        let deserialized = serde_json::from_str(&serialized).expect("must deserialize");
+        assert_eq!(input, deserialized);
+    }
+
+    #[test]
+    fn serde_deserialization() {
+        #[derive(Debug, Deserialize, PartialEq)]
+        struct Test {
+            locator: Locator,
+        }
+
+        let input = r#"{ "locator": "custom+1/foo$bar" }"#;
+        let expected = Locator::builder()
+            .fetcher(Fetcher::Custom)
+            .project("foo")
+            .revision("bar")
+            .org_id(1)
+            .build();
+        let expected = Test { locator: expected };
+
+        let deserialized = serde_json::from_str(input).expect("must deserialize");
+        assert_eq!(expected, deserialized, "deserialize {input}");
+    }
+
+    #[test]
+    fn demotes() {
+        let input = Locator::builder()
+            .fetcher(Fetcher::Custom)
+            .project("foo")
+            .org_id(1)
+            .revision("abcd")
+            .build();
+
+        let expected = PackageLocator::builder()
+            .fetcher(Fetcher::Custom)
+            .project("foo")
+            .org_id(1)
+            .build();
+        let demoted = input.clone().into_package();
+        assert_eq!(expected, demoted, "demote {input}");
+    }
+
+    #[test]
+    fn promotes_strict() {
```

Comment:
We should have a test for `promotes_strict()` where the revision already exists and does not get overridden